github-upload-action-secrets

```diff
@@ -46,10 +51,24 @@ def encrypt(public_key: str, secret_value: str) -> str:
 
     public_key = public.PublicKey(public_key.encode("utf-8"), encoding.Base64Encoder())
     sealed_box = public.SealedBox(public_key)
-    encrypted = sealed_box.encrypt(secret_value.encode("utf-8"))
+    encrypted = sealed_box.encrypt(secret_value)
     return b64encode(encrypted).decode("utf-8")
 
 
+def generate_ssh_key(pubkey):
+    """Generates a Ed25519 keypair and outputs public/private in OpenSSH format"""
+    key = Ed25519PrivateKey.generate()
+    public = key.public_key().public_bytes(Encoding.OpenSSH, PublicFormat.OpenSSH)
+    private = key.private_bytes(Encoding.PEM, PrivateFormat.OpenSSH, NoEncryption())
+
+    return public, encrypt(pubkey["key"], private)
```

Comment:
The [documentation](https://pynacl.readthedocs.io/en/latest/public/#nacl.public.Box.encrypt) makes a big point about the nonce being a nonce, not simply the private key. This should at least get some documentation what it does, and why this is safe.

Frankly, I'd feel a lot better about not reimplementing this, but call `ssh-keygen` to write files into tmpfs (`$XDG_RUNTIME_DIR`). There is even [ways to write to stdout](https://gist.github.com/kraftb/9918106).

Reply:
> The [documentation](https://pynacl.readthedocs.io/en/latest/public/#nacl.public.Box.encrypt) makes a big point about the nonce being a nonce, not simply the private key. This should at least get some documentation what it does, and why this is safe.

You're confusing the `.encrypt()` function on the box with our own `encrypt()` function which calls it like this:

```python
    encrypted = sealed_box.encrypt(secret_value)
```

ie: without providing a nonce.

In any case, this needs to be a secure operation because the public-key that you download from GitHub is the same each time.

> Frankly, I'd feel a lot better about not reimplementing this, but call `ssh-keygen` to write files into tmpfs (`$XDG_RUNTIME_DIR`). There is even [ways to write to stdout](https://gist.github.com/kraftb/9918106).

I really strongly disagree with this.  Those are some awful ugly hacks you link.  The only one that I might even be vaguely inclined to implement is the openssl one...

What's your exact concern here?  Insufficient entropy?  Degenerate keys?  Sidechannel leaks?

Reply:
Whoops! Sorry about mixing up the two encrypt(), of course you are right.

> What's your exact concern here?

Mostly that it wouldn't be the first case of someone being burned by reimplementing crypto, i.e. standard paranoia. I trust that ssh-keygen knows what it needs to do, I am not concerned about writing the key to tmpfs, and everything else I need to properly understand. :grin: So no concrete "this is unsafe".

Reply:
Replaced with "shell abuse" variant from below.